### ISSUE TITLE
Address is already in used 

### ISSUE THREAD

**Issue by Albert950207** [Unity-Technologies/Unity-Robotics-Hub#230]
Hello

In the part 2 and part3, I get an error reported by unity. The error is:
![Screenshot from 2021-05-21 13-30-45](https://user-images.githubusercontent.com/59747981/119176634-58c2dd80-ba39-11eb-86dc-a04ab0728bf7.png)
Address is already in used 
I run ROS in docker according to the part0 and my unity version is 2020.3.6f1 and I also try it in different versions but all get the same error.
Here is my IP address setting:
![Screenshot from 2021-05-21 13-37-40](https://user-images.githubusercontent.com/59747981/119178923-51e99a00-ba3c-11eb-95ac-cd2c94de9ada.png)

When I was at part 2 and got this error, I can still get the data from docker as the part shows if I turn off error pause even there is still the error "Address is already in used":
![Screenshot from 2021-05-21 16-26-33](https://user-images.githubusercontent.com/59747981/119206623-a5bea800-ba69-11eb-8797-b0ba3ad78a28.png)
![Screenshot from 2021-05-21 16-26-42](https://user-images.githubusercontent.com/59747981/119206646-b5d68780-ba69-11eb-9b3e-d87674903b5b.png)

However, when I was at part 3, after I played the game and pressed the publisher, there was another error except for "Address is already in used":
![Screenshot from 2021-05-21 21-47-12](https://user-images.githubusercontent.com/59747981/119211127-4cad3f00-ba7e-11eb-9ec5-43979d4b9a5f.png)
and here is docker after pressed publish:
![Screenshot from 2021-05-21 21-47-05](https://user-images.githubusercontent.com/59747981/119211139-63539600-ba7e-11eb-8b41-c529442cf596.png)

- Unity Version: 2020.3.6f1
- Unity machine OS + version: Ubuntu 20.04
- ROS–Unity communication: Docker

**Comment by hyounesy** [comment]
Hi @Albert950207 , I think the issue might be due to the `Override Unity IP Address` in the Unity ROS Settings not being set to the IP Address of the unity machine. Can you please replace `127.0.0.1` with the actual IP of your unity machine and try again.

**Comment by Albert950207** [comment]
Hi@hyounesy  Thank you for your solution. I tried to change the override unity IP but every time after I play the game, the IP will be reset to 127.0.0.1. What should I do ?

**Comment by hyounesy** [comment]
Could you please try closing the ROS Settings window after you set the IP Address, and then in the UnityEditor window run File -> Save Project.

**Comment by Albert950207** [comment]
@hyounesy  Thanks!  I replace the override unity IP address with  actual IP of my unity machine but it doesn't work. The errors in unity and dork are same as the previous override IP.

**Comment by Albert950207** [comment]
Does anyone know what leads to this error?

**Comment by hyounesy** [comment]
Apologies for the late response. Could you please verify the following:
- The Unity IP address you use is your "local" IP address (usually starts with 192.xxx.xxx.xxx) and not your external (public) IP assigned by the internet service provider.
- In the ROS window, the IP that is shown after "ROS-Unity Handshake received, will connect to 192.xxx.xxx.xxx:5005" matches your local Unity IP.
- In the Unity Editor, after you play the scene, click on the "View service status" checkbox and the IPs shown under MUnityHandshakeRequest and MUnityHandshakeResponse match your local Unity IP.

**Comment by sdiao** [comment]
Hi @Albert950207 , We just released several ROSConnection improvements in [ROS-TCP-Connector v0.4.0](https://github.com/Unity-Technologies/ROS-TCP-Connector/tree/v0.4.0) . We have simplified the ROS Setting window in Unity Editor, removed "Override Unity IP", and removed the Handshake. I would recommend you to check out the latest features with the pick and place tutorial in[ Unity-Robotics-Hub](https://github.com/Unity-Technologies/Unity-Robotics-Hub/tree/v0.4.0). Please let us know if there is anything else we can help with. Thank you.

**Comment by Albert950207** [comment]
Hi @hyounesy, thanks for your replying  
I have tried the way and here is my ROS setting and my local IP address 
1. ![Screenshot from 2021-06-04 18-18-12](https://user-images.githubusercontent.com/59747981/120875197-a468a180-c578-11eb-98e9-10a6347b8465.png)
2. In docker:
![Screenshot from 2021-06-04 21-21-06](https://user-images.githubusercontent.com/59747981/120875628-d1b64f00-c57a-11eb-9c49-ad9d1ac3b8dc.png)

![Screenshot from 2021-06-04 18-47-13](https://user-images.githubusercontent.com/59747981/120876107-e85da580-c57c-11eb-97f4-73db88432f22.png)

3. Service status after play
![Screenshot from 2021-06-04 18-31-23](https://user-images.githubusercontent.com/59747981/120876045-9b79cf00-c57c-11eb-9938-c3b8b825263e.png)

![Screenshot from 2021-06-04 18-46-40](https://user-images.githubusercontent.com/59747981/120876092-d4b23f00-c57c-11eb-9200-81e99d1f98c8.png)

![Screenshot from 2021-06-04 18-46-52](https://user-images.githubusercontent.com/59747981/120876103-df6cd400-c57c-11eb-8028-40e9c11d8d7b.png)



but still get error: address is already in used

**Comment by Albert950207** [comment]
Hi@sdiao, thank you for letting me know there is a new version of ROS. I will try it !

**Comment by Albert950207** [comment]
Hi @sdiao , I just checked out the new ROS-TCP connector. I update my order one in this project with the new connector by package manager in Unity. However, I got these errors.

![Screenshot from 2021-06-14 15-15-47](https://user-images.githubusercontent.com/59747981/121947129-b0482680-cd23-11eb-8201-207db891e61d.png)

Could you please help me ? Is there any instructions about how to use it in pick and place project ?

Thank you !

**Comment by sdiao** [comment]
Hi @Albert950207 , the compile error was due to an interface change in `ROSConnection`. fields including `overrideUnityIP`, `unityPort`, `awaitDataMaxRetries`,  and `awaitDataSleepSeconds` are deprecated. 
We would recommend you get the latest code from the main branch in the Unity-Robotics-Hub, where we have updated the `Demo.cs` script to use the updated ROSConnection. Please see: https://github.com/Unity-Technologies/Unity-Robotics-Hub/blob/main/tutorials/pick_and_place/PickAndPlaceProject/Assets/DemoScripts/Demo.cs#L165-L172 which shows the method `CreateRosConnection` should be:
```private void CreateRosConnection()
    {
        // Create RosConnect
        GameObject rosConnect = new GameObject(rosConnectName);
        rosConnection = rosConnect.AddComponent<ROSConnection>();
        rosConnection.RosIPAddress = hostIP;
        rosConnection.RosPort = hostPort;
    }
    ```

Let us know if the tutorial works after checking out the latest code on main branch.

**Comment by Albert950207** [comment]
Hi, @sdiao  I will the latest code on main branch. I didn't get the " address is already in used " error anymore but the arm didn't move and I still got the error "no trajectory returned from movie service"

**Comment by Albert950207** [comment]
*tried

**Comment by Albert950207** [comment]
This is error showed by my docker container
![Screenshot from 2021-07-01 11-34-52](https://user-images.githubusercontent.com/59747981/124157040-51461980-da66-11eb-8a25-dbfd441498c3.png)

These are information provided by Unity
![Screenshot from 2021-07-01 11-41-23](https://user-images.githubusercontent.com/59747981/124157200-80f52180-da66-11eb-8a78-31eb188139b6.png)
![Screenshot from 2021-07-01 11-41-50](https://user-images.githubusercontent.com/59747981/124157211-83f01200-da66-11eb-94d9-fbeaa04fe3af.png)
![Screenshot from 2021-07-01 11-42-13](https://user-images.githubusercontent.com/59747981/124157222-86eb0280-da66-11eb-8565-58628289acec.png)

I can run part 2 and received the positions.

**Comment by sdiao** [comment]
We are currently investigating the issue and will keep you posted. Just to confirm, is the error from the quick demo tutorial: https://github.com/Unity-Technologies/Unity-Robotics-Hub/blob/main/tutorials/pick_and_place/quick_demo.md?

**Comment by Albert950207** [comment]
@sdiao  Yes, it is from the pick and place demo and I rebuild my docker image with latest code but I was having the old one when I rebuild. I am not sure if it will cover the old one.

**Comment by sdiao** [comment]
@Albert950207 , you can use `docker container ls` to list the containers, and use `docker rm [OPTIONS] CONTAINER` to remove the pick and place container, then rebuild the docker image with `cd /PATH/TO/Unity-Robotics-Hub/tutorials/pick_and_place &&
git submodule update --init --recursive &&
docker build -t unity-robotics:pick-and-place -f docker/Dockerfile .`

It seems some part of the tutorial is out of sync with the latest main branch for your local environment. I would recommend that you delete the Unity-Robotics-Hub repo first and download it again, re-build docker images and set up the Unity project following the tutorial. 
Please make sure to use the git submodule when you clone the repo `git clone --recurse-submodules https://github.com/Unity-Technologies/Unity-Robotics-Hub.git`. 

Hope it helps. Let us know if removing&rebuilding docker container or starting the tutorial from scratch works. Thank you.

**Comment by Albert950207** [comment]
@sdiao  Hi, thank you for your suggestions. The things are that when I exit the docker container, I can't see it anymore by using 'docker container ls' so I just download the latest repository into a new directory and rebuild the docker image according to the tutorials. However, it still doesn't work and same errors are returned.


![Screenshot from 2021-07-09 17-55-20](https://user-images.githubusercontent.com/59747981/125142852-c066e200-e0e6-11eb-8a1e-aecccc1faed4.png)


![Screenshot from 2021-07-09 17-54-00](https://user-images.githubusercontent.com/59747981/125142856-c3fa6900-e0e6-11eb-9960-f033215d86ff.png)

**Comment by sdiao** [comment]
Hi @Albert950207 , Let's set up a 30 minutes meeting to troubleshoot the issue. Could you send 3 of your availabilities to unity-robotics@unity3d.com? The Robotics team is in Pacific Daylight Timezone (PDT). Thanks!

**Comment by Albert950207** [comment]
@sdiao Yes, sure! I have sent it just right now

**Comment by sdiao** [comment]
@Albert950207 , I just wanted to follow up and see if you are able to get the tutorial part 2 and part3 working after the meeting. Let us know. Thanks.

**Comment by sdiao** [comment]
I am going to close the issue now since we have been able to run the quick demo in the meeting. Please feel free to reach out if you have any questions in the future. Thanks!

**Comment by NathanielRose** [comment]
Following up on this thread, It looks like I receive an Identical issue when I switch between scenes  with an established ROS connection. What were the outcomes of the meeting that mitigated this error @sdiao? I have an open issue #314  that seeks support on this.

**Comment by sdiao** [comment]
Hi @NathanielRose , for the meeting, we tried the Pick and Place [quick demo ](https://github.com/Unity-Technologies/Unity-Robotics-Hub/blob/v0.4.0/tutorials/pick_and_place/quick_demo.md) , and verified that demo project works for the customer. 
We suspected the customer's local environment did not have a clean setup. Starting the tutorial from scratch solved the connection issue. 
Let's move the discussion to the new issue that you opened.
